### DISCUSSION TITLE
Question: Convert error occurs with null-conditional, null-forgiving, and extension method when update VS to 16.8

### DISCUSSION THREAD

**Discussion by tzengshinfu** [dotnet/csharplang#4129] [Q&A] [answered]
Hi, folks:

I write a extension method that convert null to default string like this:
```csharp
public static string ToStringOrDefault(this object obj, string defaultValue) {
    //When obj is null then returns defaultValue.
}
```

And I use the extension like this:
```csharp
using System.Xml.Linq;

XElement rootNode = XDocument.Parse("some XML text...").Root;
bool someNode = rootNode.Element("some Node name...")?.Value!.ToStringOrDefault("N").Contains("Y");
```

Everything is ok in VS 16.7 + .NET Core 3.1.
When I update VS to 16.8, I get errors like this:
_CS0266 - Cannot implicitly convert type 'bool?' to 'bool'. An explicit conversion exists (are you missing a cast?)_

Could someone tell me what happend?

**Comment by ufcpp** [comment, marked as answer]
https://github.com/dotnet/csharplang/issues/3393

**Reply by tzengshinfu** [reply]
Hi, Iwanaga san:

Thanks for your reply.
I added parentheses to rootNode.Element("some Node name...")?.Value to resolve these errors like this:
```csharp
bool someNode = (rootNode.Element("some Node name...")?.Value).ToStringOrDefault("N").Contains("Y");
```

**Comment by RikkiGibson** [comment]
If your extension method accepts a null `this` argument, consider changing the parameter to `this object? obj`. If you do that, nullability analysis will just do the right thing--it'll see that the `this` argument could be null, but that the extension method accepts a null `this` argument, so everything is fine.

**Reply by tzengshinfu** [reply]
Hi, Riki:

Thanks for your reply.
I tried to change object to object?, but it still doesn't call the extension method and returns null directly.

**Reply by Joe4evr** [reply]
Because the purpose of `?.` is to short-circuit and *not* call the right-hand side if the left-hand side is `null`.

**Reply by RikkiGibson** [reply]
That's right, my suggestion doesn't solve the problem you're experiencing, but it does make the signature of your method more accurate :)

**Reply by tzengshinfu** [reply]
> That's right, my suggestion doesn't solve the problem you're experiencing, but it does make the signature of your method more accurate :)

Thanks, Rikki!

**Reply by tzengshinfu** [reply]
> Because the purpose of `?.` is to short-circuit and _not_ call the right-hand side if the left-hand side is `null`.

Hi, Joe4evr:

Thanks for your explanation.
It's a shame that null-forgiving operator(!) is not working now.
I like
```csharp
rootNode.Element("some Node name...")?.Value!.ToStringOrDefault("N")
```
more than
```csharp
(rootNode.Element("some Node name...")?.Value).ToStringOrDefault("N")
```
